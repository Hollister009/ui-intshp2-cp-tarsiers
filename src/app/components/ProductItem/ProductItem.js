import React, { Component } from 'react';
import { DisplayFront, DisplayDetails } from './ProductDisplay';
import HttpService from '../../../utils/http.service';
import appConfig from '../../../config/appConfig';
import styles from './ProductItem.module.scss';
import '../WishListContainer/WishListContainer.scss';

class ProductItem extends Component {
  constructor(props) {
    super(props);
    this.state = { showDetails: false, activeItem: false };
    this.itemRef = React.createRef();
  }

  onHoverItem = () =>
    this.setState(state => ({ activeItem: !state.activeItem }));

  showFront = () => this.setState({ showDetails: false });

  showDetails = () => this.setState({ showDetails: true });

  clickHandle = (e, id) => {
    e.preventDefault();

    HttpService.post(appConfig.apiResources.wishList, { productId: id }).then(
      res => console.log('res', res)
    );
  };

  render() {
    const { data, extended } = this.props;
    const { showDetails, activeItem } = this.state;
    const cardView = showDetails ? (
      <DisplayDetails {...data} clickHandler={this.clickHandle} />
    ) : (
      <DisplayFront {...data} />
    );

    return extended ? (
      <div
        ref={this.itemRef}
        className={styles.product_card}
        onMouseEnter={this.showDetails}
        onMouseLeave={this.showFront}
      >
        {cardView}
      </div>
    ) : (
      <div
        ref={this.itemRef}
        className="item"
        onMouseEnter={this.onHoverItem}
        onMouseLeave={this.onHoverItem}
      >
        <div className="item-pic">
          <img src={data.src} alt="pic" />
        </div>
        <div className="item-info">
          <b>{data.title}</b>
          {activeItem ? (
            <div>
              <button type="button" className="add-to-card">
                <i className="fas fa-shopping-cart" />
                add to cart
              </button>
            </div>
          ) : (
            <div>
              <div className="stars">
                <i className="fa fa-star" />
                <i className="fa fa-star" />
                <i className="fa fa-star" />
              </div>
              <span className="price">{`${data.price} $`}</span>
            </div>
          )}
        </div>
      </div>
    );
  }
}

export default ProductItem;
