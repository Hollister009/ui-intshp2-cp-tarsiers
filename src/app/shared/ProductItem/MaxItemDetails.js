/* eslint-disable no-console */
import React, { Component } from 'react';
import { Flags } from 'react-feature-flags';
import { Notify } from 'react-redux-notify';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import HttpService from '../../../utils/http.service';
import appConfig from '../../../config/appConfig';
import productType from '../../../types';
import NotifyService from '../../../utils/notify.service';

const { addToWishList, removeFromWishList } = appConfig.apiResources;

const CN = 'product-item--full';

class MaxItemDetails extends Component {
  static propTypes = {
    /**
     * data - productType shape
     */
    data: productType,
    addToWishListItem: PropTypes.func.isRequired,
    removeFromWishListItem: PropTypes.func.isRequired,
    wished: PropTypes.bool
  };

  static defaultProps = { data: null, wished: false };

  state = { heartDisabled: false };

  componentDidMount() {
    const { data } = this.props;

    this.setState({ image: data.src });
  }

  addItem = id => {
    const { addToWishListItem, wished, createNotification } = this.props;

    HttpService.post(addToWishList, { productId: id })
      .then(res => {
        if (res.status === 200 && !wished) {
          addToWishListItem(id);
          console.log(`Added to the WishList: ${id}`);
        }
      })
      .catch(error => console.log(error))
      .finally(this.setState({ heartDisabled: false }));

    createNotification(NotifyService.added);
  };

  removeItem = id => {
    const { removeFromWishListItem, createNotification } = this.props;

    HttpService.post(removeFromWishList, { productId: id })
      .then(res => {
        if (res.status === 200) {
          removeFromWishListItem(id);
          console.log(`Removed from the WishList: ${id}`);
        }
      })
      .catch(error => console.log(error))
      .finally(this.setState({ heartDisabled: false }));

    createNotification(NotifyService.removed);
  };

  toggleSwatch = (e, colors) => {
    if (e.keyCode === 13 || e.type === 'click') {
      const value = e.target.innerText;

      this.setState({ image: colors[value] });
    }
  };

  swatchesFactory = (pins, items) =>
    pins.map(color => (
      <span
        key={color}
        role="button"
        tabIndex="0"
        style={{ backgroundColor: `${color}` }}
        onClick={e => this.toggleSwatch(e, items)}
        onKeyDown={e => this.toggleSwatch(e, items)}
      >
        {color}
      </span>
    ));

  toggleWishList = (e, id) => {
    const { wished } = this.props;

    e.preventDefault();
    const cb = !wished ? this.addItem : this.removeItem;

    this.setState({ heartDisabled: true }, () => {
      cb(id);
    });
  };

  cartNote = () => {
    const { createNotification } = this.props;

    createNotification(NotifyService.cart);
  };

  render() {
    const { data } = this.props;
    const { _id, title, sizes, colors, colorUrls, wished } = data;
    const { heartDisabled, image } = this.state;
    const allSizes = sizes
      .map((size, i) => (i !== 0 ? `- ${size}` : size))
      .join(' ');
    const swatches = this.swatchesFactory(colors, colorUrls);

    return (
      data && (
        <React.Fragment>
          <Link to={`/products/${_id}`}>
            <img className={`${CN}__img-small`} src={image} alt={data.title} />
            <h4 className="highlighted">{title}</h4>
          </Link>
          <div className={`${CN}__sizes`}>{`sizes : ${allSizes}`}</div>
          <div className={`${CN}__swatches`}>{swatches}</div>
          <hr className="separate" />
          <div className="social_buttons">
            <button type="button" title="Share with others">
              <i className="fas fa-share-alt" />
            </button>
            <Notify position={NotifyService.position.topRight} />
            <button
              type="button"
              title="Add to shopping-cart"
              onClick={this.cartNote}
            >
              <i className="fas fa-shopping-cart" />
            </button>
            <Flags authorizedFlags={[appConfig.killswitch.wishlist]}>
              <button
                type="button"
                className="btn-heart"
                onClick={e => this.toggleWishList(e, _id)}
                title="Add to wish-list"
                disabled={heartDisabled}
              >
                <i
                  className={
                    wished ? 'fas fa-heart highlighted' : 'fas fa-heart'
                  }
                />
              </button>
            </Flags>
          </div>
        </React.Fragment>
      )
    );
  }
}

export default MaxItemDetails;
