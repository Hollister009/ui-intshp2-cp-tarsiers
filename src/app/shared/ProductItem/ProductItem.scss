@import '../../../styles/variables';
@import '../../../styles/mixin';

.product-item {
  &--full {
    min-width: 270px;
    height: 365px;
    max-height: 365px;
    margin-bottom: 30px;
    padding: 30px 0;
    margin: 0 15px;
    text-align: center;
    color: $color-dark;
    background-color: $color-white;
    box-shadow: 0 0 10px rgba($color-dark, 0.2);

    &:first-child {
      margin-left: 0;
    }

    &:last-child {
      margin-right: 0;
    }

    img {
      margin: 0 auto;
    }

    h4 {
      font-size: 18px;
      margin-top: 30px;
    }

    span {
      display: inline-block;
      font-size: 18px;
      font-weight: 700;
      padding: 18px 0;
    }

    button {
      background-color: transparent;
      border: none;
      font-size: 22px;
      color: inherit;
      cursor: pointer;
      transition: color 0.3s ease;

      &:nth-child(2) {
        margin: 0 25px;
      }

      &:hover,
      &:focus {
        color: $color-main;
        outline: none;
      }
    }

    &__sizes {
      font-size: 14px;
      line-height: 1;
      margin-top: 30px;
    }

    &__img_small {
      width: 70px;
      height: 87px;
    }

    &__swatches {
      width: 70px;
      margin: 30px auto;
      @include d-flex();

      & > div {
        width: 15px;
        height: 15px;
        font-size: 0;
        border-radius: 50%;
      }
    }
  }

  &--small {
    margin: 0 15px;
    display: flex;
    min-width: 370px;
    box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
    background-color: $color-white;

    &:first-child {
      margin: 0 15px 0 0;
    }
    &:hover {
      box-shadow: 0 0 10px rgba(0, 0, 0, 0.3);
    }

    .item-pic img {
      padding: 36px;
      width: 150px;
      max-width: initial;
    }

    .item-info {
      margin: 0 auto;
      padding: 5px;
      @include d-flex(center, column);

      b {
        margin-bottom: 20px;
        justify-content: center;
      }
    }

    .stars {
      display: inline-block;
      padding: 0 20px;
    }

    .price {
      display: inline-block;
      font-family: $main-font-bold;
      color: $color-main;
      padding-right: 10px;
    }

    .add-to-card {
      border: none;
      color: $color-main;
      font-size: 16px;

      i {
        padding-right: 5px;
      }
    }
  }
}

@media screen and (max-width: $media-mobile) {
  .product-item--full {
    height: 330px;
    min-width: 240px;
    img {
      height: auto;
      width: 150px;
    }
    h4 {
      margin: 0;
    }
    &__img_small {
      padding: none;
    }
  }
}
