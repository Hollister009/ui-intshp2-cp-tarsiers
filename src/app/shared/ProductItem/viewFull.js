/* eslint-disable no-console */
import React, { Component } from 'react';
import { Flags } from 'react-feature-flags';

import HttpService from '../../../utils/http.service';
import appConfig from '../../../config/appConfig';

const { addToWishList, removeFromWishList } = appConfig.apiResources;

const CN = 'product-item--full';

const ViewFrontFull = props => {
  const { src, title, price } = props;

  return (
    <React.Fragment>
      <img className={`${CN}__img`} src={src} alt="" />
      <h4>{title}</h4>
      <span className="highlighted">{`${price} $`}</span>
    </React.Fragment>
  );
};

class ViewDetailsFull extends Component {
  state = { heartDisabled: false };

  static defaultProps = { wished: false };

  addItem = id => {
    const { addToWishListItem, wished } = this.props;

    HttpService.post(addToWishList, { productId: id })
      .then(res => {
        if (res.status === 200 && !wished) {
          addToWishListItem(id);
          console.log(`Added to the WishList: ${id}`);
        }
      })
      .catch(error => console.log(error))
      .finally(this.setState({ heartDisabled: false }));
  };

  removeItem = id => {
    const { removeFromWishListItem } = this.props;

    HttpService.post(removeFromWishList, { productId: id })
      .then(res => {
        if (res.status === 200) {
          removeFromWishListItem(id);
          console.log(`Removed from the WishList: ${id}`);
        }
      })
      .catch(error => console.log(error))
      .finally(this.setState({ heartDisabled: false }));
  };

  toggleWishList = (e, id) => {
    const { wished } = this.props;

    e.preventDefault();
    const cb = !wished ? this.addItem : this.removeItem;

    this.setState({ heartDisabled: true }, () => {
      cb(id);
    });
  };

  render() {
    const { _id, src, title, sizes, colors, wished } = this.props;
    const { heartDisabled } = this.state;
    const colorPins = colors.map(color => (
      <div key={color} style={{ backgroundColor: `${color}` }} />
    ));
    const allSizes = sizes
      .map((size, i) => (i !== 0 ? `- ${size}` : size))
      .join(' ');

    return (
      <React.Fragment>
        <img className={`${CN}__img-small`} src={src} alt="" />
        <h4 className="highlighted">{title}</h4>
        <div className={`${CN}__sizes`}>{`sizes : ${allSizes}`}</div>
        <div className={`${CN}__swatches`}>{colorPins}</div>
        <hr className="separate" />
        <div className="social_buttons">
          <button type="button" title="Share with others">
            <i className="fas fa-share-alt" />
          </button>
          <button type="button" title="Add to shopping-cart">
            <i className="fas fa-shopping-cart" />
          </button>
          <Flags authorizedFlags={[appConfig.killswitch.wishlist]}>
            <button
              type="button"
              className="btn-heart"
              onClick={e => this.toggleWishList(e, _id)}
              title="Add to wish-list"
              disabled={heartDisabled}
            >
              <i
                className={wished ? 'fas fa-heart highlighted' : 'fas fa-heart'}
              />
            </button>
          </Flags>
        </div>
      </React.Fragment>
    );
  }
}

export { ViewFrontFull, ViewDetailsFull };
