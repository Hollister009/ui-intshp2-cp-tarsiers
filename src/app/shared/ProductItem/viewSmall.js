import React from 'react';

const CN = 'product-item--small';

const ViewCartSmall = props => {
  const { title } = props;

  return (
    <React.Fragment>
      <div className={`${CN}__info`}>
        <h4>{title}</h4>
        <button type="button" className="add-to-card">
          <i className="fas fa-shopping-cart" />
          add to cart
        </button>
      </div>
    </React.Fragment>
  );
};

const ViewInfoSmall = props => {
  const { title, price } = props;

  return (
    <React.Fragment>
      <div className={`${CN}__info`}>
        <h4>{title}</h4>
        <div className="info-group">
          <div className="rating">
            <i className="fa fa-star" />
            <i className="fa fa-star" />
            <i className="fa fa-star" />
          </div>
          <span className="highlighted price">{`${price} $`}</span>
        </div>
      </div>
    </React.Fragment>
  );
};

export { ViewCartSmall, ViewInfoSmall };
