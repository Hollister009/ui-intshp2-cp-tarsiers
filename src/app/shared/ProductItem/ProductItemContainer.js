import { connect } from 'react-redux';
import { createNotification } from 'react-redux-notify';
import { addToWishList, removeFromWishList } from '../../actions';
import ProductItem from './ProductItem';

const mapStateToProps = state => ({ wishlist: state.wishlist.wishlist });
const mapDispatchToProps = dispatch => ({
  addToWishListItem: data => dispatch(addToWishList(data)),
  removeFromWishListItem: data => dispatch(removeFromWishList(data)),
  createNotification: config => dispatch(createNotification(config))
});

const ProductItemContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(ProductItem);

export default ProductItemContainer;
