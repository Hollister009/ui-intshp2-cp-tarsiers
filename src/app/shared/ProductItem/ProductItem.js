/* eslint-disable object-curly-newline */
import React, { Component } from 'react';
import { bool } from 'prop-types';

import productType from '../../../types';
import { ViewFrontFull, ViewDetailsFull } from './viewFull';
import { ViewCartSmall, ViewInfoSmall } from './viewSmall';
import './ProductItem.scss';

const CN = 'product-item';

class ProductItem extends Component {
  static propTypes = {
    extended: bool,
    isAddedtoWishList: bool,

    /**
     * data - productType shape
     */
    data: productType
  };

  static defaultProps = {
    extended: false,
    isAddedtoWishList: false,
    data: {}
  };

  state = { showDetails: false };

  showFront = () => this.setState({ showDetails: false });

  showDetails = () => this.setState({ showDetails: true });

  render() {
    const { showDetails } = this.state;
    const {
      data,
      extended,
      isAddedtoWishList,
      addToWishListItem,
      removeFromWishListItem
    } = this.props;
    const { src, title, price } = data;

    return extended ? (
      <div
        className={`${CN} ${CN}--full`}
        onMouseEnter={this.showDetails}
        onMouseLeave={this.showFront}
      >
        {showDetails ? (
          <ViewDetailsFull
            {...data}
            addToWishListItem={addToWishListItem}
            removeFromWishListItem={removeFromWishListItem}
            wished={isAddedtoWishList}
          />
        ) : (
          <ViewFrontFull src={src} title={title} price={price} />
        )}
      </div>
    ) : (
      <div
        className={`${CN} ${CN}--small`}
        onMouseEnter={this.showDetails}
        onMouseLeave={this.showFront}
      >
        <img src={src} alt="" />
        {showDetails ? (
          <ViewCartSmall title={title} />
        ) : (
          <ViewInfoSmall title={title} price={price} />
        )}
      </div>
    );
  }
}

export default ProductItem;
