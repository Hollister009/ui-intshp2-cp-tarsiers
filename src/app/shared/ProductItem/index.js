import React, { Component } from 'react';

import HttpService from '../../../utils/http.service';
import appConfig from '../../../config/appConfig';
import { ViewFrontFull, ViewDetailsFull } from './viewFull';
import { ViewCartSmall, ViewInfoSmall } from './viewSmall';
import './ProductItem.scss';

class ProductItem extends Component {
  constructor(props) {
    super(props);
    this.state = { showDetails: false };
    this.itemRef = React.createRef();
  }

  showFront = () => this.setState({ showDetails: false });

  showDetails = () => this.setState({ showDetails: true });

  clickHandle = (e, id) => {
    e.preventDefault();

    HttpService.post(appConfig.apiResources.wishList, { productId: id }).then(
      res => console.log('res', res)
    );
  };

  render() {
    const { showDetails } = this.state;
    const { data, extended } = this.props;

    return extended ? (
      <div
        ref={this.itemRef}
        className="product-item--full"
        onMouseEnter={this.showDetails}
        onMouseLeave={this.showFront}
      >
        {showDetails ? (
          <ViewDetailsFull {...data} clickHandler={this.clickHandle} />
        ) : (
          <ViewFrontFull {...data} />
        )}
      </div>
    ) : (
      <div
        ref={this.itemRef}
        className="product-item--small"
        onMouseEnter={this.showDetails}
        onMouseLeave={this.showFront}
      >
        {showDetails ? (
          <ViewCartSmall {...data} />
        ) : (
          <ViewInfoSmall {...data} />
        )}
      </div>
    );
  }
}

export default ProductItem;
