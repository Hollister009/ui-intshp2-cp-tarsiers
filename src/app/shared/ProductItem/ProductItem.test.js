import React from 'react';
import { shallow, mount } from 'enzyme';

import ProductItem from './ProductItem';

const products = require('../../../mocks/products.json');

xdescribe('<ProductItem />', () => {
  let props;
  let extended;

  beforeEach(() => {
    extended = true;
    props = { data: products[1] };
  });

  it('should match its snapshot', () => {
    const shallowWrapper = shallow(<ProductItem {...props} />);

    expect(shallowWrapper).toMatchSnapshot();
  });

  xdescribe('<viewSmall />', () => {
    it('should be able to render <ViewCartSmall />', () => {
      const wrapper = shallow(<ProductItem {...props} />);
      const instance = wrapper.instance();

      instance.showFront();
      expect(wrapper.state('showDetails')).toBe(false);
    });

    it('should be able to render <ViewInfoSmall />', () => {
      const wrapper = shallow(<ProductItem {...props} />);
      const instance = wrapper.instance();

      instance.showDetails();
      expect(wrapper.state('showDetails')).toBe(true);
    });

    it('should render <ViewInfoSmall /> if showDetails is true', () => {
      const wrapper = mount(<ProductItem {...props} />);

      wrapper.setState({ showDetails: true });
      expect(wrapper.state('showDetails')).toBe(true);
    });
  });

  xdescribe('<viewFull>', () => {
    it('should be able to render <ViewFrontFull />', () => {
      const wrapper = shallow(<ProductItem {...props} extended={extended} />);
      const instance = wrapper.instance();

      instance.showFront();
      expect(wrapper.state('showDetails')).toBe(false);
    });

    it('should be able to render <ViewDetailsFull />', () => {
      const wrapper = shallow(<ProductItem {...props} extended={extended} />);
      const instance = wrapper.instance();

      instance.showDetails();
      expect(wrapper.state('showDetails')).toBe(true);
    });

    it('should render <ViewDetailsFull /> if showDetails is true', () => {
      const wrapper = mount(<ProductItem {...props} extended={extended} />);

      wrapper.setState({ showDetails: true });
      expect(wrapper.state('showDetails')).toBe(true);
    });

    it('should handle click on wishlist button', () => {
      const wrapper = mount(<ProductItem {...props} extended={extended} />);

      wrapper.setProps({ isAddedtoWishList: true });
      wrapper.setState({ showDetails: true });
      wrapper.update();
      wrapper.find('.btn-heart').simulate('click');
    });
  });
});
