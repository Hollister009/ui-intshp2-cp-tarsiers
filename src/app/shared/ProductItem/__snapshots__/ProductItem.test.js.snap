// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductItem /> should match its snapshot 1`] = `
<div
  className="product-item product-item--small"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <img
    alt=""
  />
  <ViewInfoSmall />
</div>
`;
